File: src/Core/Data/EntitySet.cs
using System.Collections;

namespace AppBoxCore;

/// <summary>
/// 实体的EntitySet成员，添加项时设置关联的EntityRef成员, 记录移除项的列表
/// </summary>
public sealed class EntitySet<T> : IList<T>, IBinSerializable where T : Entity, new()
{
    /// <summary>
    /// 创建EntitySet实例
    /// </summary>
    /// <param name="entityRefSetter">添加或移除时设置相应的EntityRef引用 eg: (t,toNull) => t.Order = toNull ? null : this </param>
    public EntitySet(Action<T, bool> entityRefSetter)
    {
        _entityRefSetter = entityRefSetter;
        _list = [];
    }

    private readonly Action<T, bool> _entityRefSetter;
    private readonly List<T> _list;
    private List<T>? _removed; //已删除的已持久化的实例

    /// <summary>
    /// 已移除的项目
    /// </summary>
    public IList<T>? RemovedList => _removed;

    public void Add(T item)
    {
        _entityRefSetter(item, false);
        _list.Add(item);
    }

    public void Insert(int index, T item)
    {
        _entityRefSetter(item, false);
        _list.Insert(index, item);
    }

    public bool Remove(T item)
    {
        var res = _list.Remove(item);
        if (res)
            RemoveInternal(item);
        return res;
    }

    public void RemoveAt(int index)
    {
        RemoveInternal(_list[index]);
        _list.RemoveAt(index);
    }

    public void Clear()
    {
        foreach (var item in _list)
        {
            RemoveInternal(item);
        }

        _list.Clear();
    }

    private void RemoveInternal(T item)
    {
        if (item is DbEntity dbEntity && dbEntity.PersistentState != PersistentState.Detached)
        {
            _removed ??= new List<T>();
            _removed.Add(item);
        }

        _entityRefSetter(item, true);
    }

    public T this[int index]
    {
        get => _list[index];
        [PixUI.TSIndexerSetToMethod]
        set
        {
            RemoveInternal(_list[index]);
            _entityRefSetter(value, false);
            _list[index] = value;
        }
    }

    public int Count => _list.Count;
    public bool IsReadOnly => false;

    public int IndexOf(T item) => _list.IndexOf(item);
    public bool Contains(T item) => _list.Contains(item);
    public void CopyTo(T[] array, int arrayIndex) => _list.CopyTo(array, arrayIndex);
    public IEnumerator<T> GetEnumerator() => _list.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    /// <summary>
    /// Only for DbStore.SaveEntitySet()
    /// </summary>
    internal void ClearRemoved() => _removed = null;

    #region ====Serialization====

    void IBinSerializable.WriteTo(IOutputStream ws)
    {
        ws.WriteVariant(_list.Count);
        foreach (var item in _list)
        {
            ws.Serialize(item);
        }

        ws.WriteVariant(_removed?.Count ?? 0);
        if (_removed != null)
        {
            foreach (var item in _removed)
            {
                ws.Serialize(item);
            }
        }
    }

    void IBinSerializable.ReadFrom(IInputStream rs)
    {
        var count = rs.ReadVariant();
        for (var i = 0; i < count; i++)
        {
            _list.Add(rs.DeserializeEntity(() => new T())!); //直接加入列表
        }

        count = rs.ReadVariant();
        if (count > 0)
        {
            _removed = new List<T>(count);
            for (var i = 0; i < count; i++)
            {
                _removed.Add(rs.DeserializeEntity(() => new T())!);
            }
        }
    }

    #endregion
}

public static class EntitySetExtensions
{
    public static void AcceptChanges<T>(this EntitySet<T> entitySet) where T : DbEntity, new()
    {
        if (entitySet.RemovedList != null)
        {
            for (var i = 0; i < entitySet.RemovedList.Count; i++)
            {
                if (entitySet.RemovedList[i].PersistentState !=
                    PersistentState.Detached /*Maybe already changed by DbStore*/)
                    entitySet.RemovedList[i].AcceptChanges();
            }

            entitySet.ClearRemoved();
        }

        for (var i = 0; i < entitySet.Count; i++)
        {
            if (entitySet[i].PersistentState != PersistentState.Unchanged)
                entitySet[i].AcceptChanges();
        }
    }
}

File: src/Dev.Web/src/AppBoxCore/Data/EntitySet.ts
import {DbEntity, Entity, IBinSerializable, IInputStream, IOutputStream, PersistentState} from "@/AppBoxCore";
import {Action2, Func1, List} from "@/System";

export class EntitySet<T extends Entity> extends List<T> implements IBinSerializable {

    constructor(entityRefSetter: Action2<T, boolean>, creator: Func1<T>) {
        super();
        this._entityRefSetter = entityRefSetter;
        this._creator = creator;
    }

    private readonly _entityRefSetter: Action2<T, boolean>;
    private readonly _creator: Func1<T>;
    private _removed: List<T> | null = null;

    public Add(item: T): void {
        this._entityRefSetter(item, false);
        super.Add(item);
    }

    public Remove(item: T): boolean {
        let res = super.Remove(item);
        if (res)
            this.RemoveInternal(item);
        return res;
    }

    private RemoveInternal(item: T) {
        if (item instanceof DbEntity && item.PersistentState != PersistentState.Detached) {
            this._removed ??= new List<T>();
            this._removed.Add(item);
        }

        this._entityRefSetter(item, true);
    }

    //region ====Serialization====
    WriteTo(bs: IOutputStream): void {
        bs.WriteVariant(this.length);
        for (const item of this) {
            (<any>bs).Serialize(item);
        }
        bs.WriteVariant(this._removed?.length ?? 0);
        if (this._removed != null) {
            for (const item of this._removed) {
                (<any>bs).Serialize(item);
            }
        }
    }

    ReadFrom(bs: IInputStream): void {
        let count = bs.ReadVariant();
        for (let i = 0; i < count; i++) {
            super.Add((<any>bs).DeserializeEntity(/*this._creator*/ null));
        }

        count = bs.ReadVariant();
        if (count > 0) {
            this._removed = new List<T>();
            for (let i = 0; i < count; i++) {
                this._removed.Add((<any>bs).DeserializeEntity(/*this._creator*/ null));
            }
        }
    }

    //endregion
}
